tests/test_fog.py

```diff
@@ -0,0 +1,244 @@
+    fog = HexaryTrieFog()
+
+    empty_prefix = ()
+    branched = fog.explore(empty_prefix, ((1,), (5,)))
+    assert not branched.is_complete
+
+    # complete all sub-segments at once
+    completed = branched.mark_all_complete(((1,), (5,)))
+    assert completed.is_complete
+
+
+def test_trie_fog_composition_equality():
+    fog = HexaryTrieFog()
+
+    empty_prefix = ()
+    single_exploration = fog.explore(empty_prefix, ((9, 9, 9), ))
+
+    half_explore = fog.explore(empty_prefix, ((9, ),))
+    full_explore = half_explore.explore((9, ), ((9, 9), ))
+
+    assert single_exploration == full_explore
+
+
+def test_trie_fog_immutability():
+    fog = HexaryTrieFog()
+
+    fog1 = fog.explore((), ((1, ), (2, )))
+
+    fog2 = fog1.explore((1, ), ((3, ), ))
+
+    assert fog.nearest_unknown(()) == ()
+    assert fog1.nearest_unknown(()) == (1,)
+    assert fog2.nearest_unknown(()) == (1, 3)
+
+    assert fog != fog1
+    assert fog1 != fog2
+    assert fog != fog2
+
+
+@pytest.mark.parametrize('sub_segments', (
+    [(1, 2), (1, 2, 3, 4)],
+    [(1, 2), (1, 2)],
+))
+def test_trie_fog_explore_invalid(sub_segments):
+    """
+    Cannot explore with a sub_segment that is a child of another sub_segment,
+    or a duplicate
+    """
+    fog = HexaryTrieFog()
+    with pytest.raises(ValidationError):
+        fog.explore((), sub_segments)
+
+
+def test_trie_fog_nearest_unknown():
+    fog = HexaryTrieFog()
+
+    empty_prefix = ()
+    assert fog.nearest_unknown((1, 2, 3)) == empty_prefix
+
+    branched = fog.explore(empty_prefix, ((1, 1), (5, 5)))
+
+    # Test shallower
+    assert branched.nearest_unknown((0,)) == (1, 1)
+    assert branched.nearest_unknown((1,)) == (1, 1)
+    assert branched.nearest_unknown((2,)) == (1, 1)
+    assert branched.nearest_unknown((4,)) == (5, 5)
+    assert branched.nearest_unknown((5,)) == (5, 5)
+    assert branched.nearest_unknown((6,)) == (5, 5)
+
+    # Test same level
+    assert branched.nearest_unknown((0, 9)) == (1, 1)
+    assert branched.nearest_unknown((1, 1)) == (1, 1)
+    assert branched.nearest_unknown((2, 1)) == (1, 1)
+    assert branched.nearest_unknown((3, 2)) == (1, 1)
+    assert branched.nearest_unknown((3, 3)) == (5, 5)
+    assert branched.nearest_unknown((4, 9)) == (5, 5)
+    assert branched.nearest_unknown((5, 5)) == (5, 5)
+    assert branched.nearest_unknown((6, 1)) == (5, 5)
+
+    # Test deeper
+    assert branched.nearest_unknown((0, 9, 9)) == (1, 1)
+    assert branched.nearest_unknown((1, 1, 0)) == (1, 1)
+    assert branched.nearest_unknown((2, 1, 1)) == (1, 1)
+    assert branched.nearest_unknown((4, 9, 9)) == (5, 5)
+    assert branched.nearest_unknown((5, 5, 0)) == (5, 5)
+    assert branched.nearest_unknown((6, 1, 1)) == (5, 5)
+
+
+def test_trie_fog_nearest_unknown_fully_explored():
+    fog = HexaryTrieFog()
+    empty_prefix = ()
+    fully_explored = fog.explore(empty_prefix, ())
+
+    with pytest.raises(IndexError):
```

Comment:
Looking at this test, I suspect we might be better served with a custom exception.  When I browsed the code it didn't stand out, but now I'm thinking about call sites where we'll have to catch this and `IndexError` doesn't make much sense to me as opposed to something more naturally descriptive like `FullyExplored`.